hack/check.sh

```diff
@@ -34,4 +34,19 @@ if [[ "$unformatted_files" ]]; then
   exit 1
 fi
 
-echo "All checks successful"
\ No newline at end of file
+echo "Checking Go version"
+while IFS= read -r line
+do
+  if [[ $line =~ ^go.*$ ]]; then
+    if [[ $line =~ ^go\ [0-9]+\.[0-9]+\.0$ ]]; then
```

Comment:
Also there is an easy way to get the go version from go mod
```bash
go list -f {{.GoVersion}} -m
```

Reply:
Thanks for the review @unmarshall!

**Why it is required to use `x.y.0`**

gardener/gardener maintainers have decided to use `x.y.0`, in line with kubernetes/kubernetes and controller-runtime.

The reason being that bumping direct dependencies would also change go.mod Go minor/patch versions and thus the change in Go versions are interspersed with dependency upgrades which they did not want.

See https://github.com/gardener/gardener/pull/10841#issuecomment-2473115882, and especially https://github.com/gardener/gardener/pull/10841#issuecomment-2473125245 where they asked us to release a patch which changes the Go version of etcd-druid just so we are consumable by them.

Fixes in the stdlib, runtime, etc are still made use of since the latest Go command release is used.

****

Thanks for the suggestion regarding `go list...` for the go version from go.mod! Will make use of it.

Reply:
Ah right - the following might answer your question @unmarshall:

> Afaik, go x.y breaks forward compatibility ([ref](https://go.dev/blog/toolchain#forward)) starting with Go 1.21.0 too, because Go is not able to identify the toolchain in this case. Since Go 1.21 the toolchain name changed from go1.20 to go1.21.0 ([ref](https://go.dev/doc/toolchain)).

The above is https://github.com/gardener/gardener/pull/9564#issuecomment-2049233929.

Due to this, having the patch version as well would be the right way to go.